src/boot/efi/stub.c

```diff
@@ -350,22 +445,37 @@ static EFI_STATUS cmdline_append_and_measure_addons(
                         continue;
                 }
 
-                _cleanup_free_ char16_t *tmp = TAKE_PTR(buffer),
-                                        *extra16 = xstrn8_to_16((char *)loaded_addon->ImageBase + addrs[UNIFIED_SECTION_CMDLINE],
-                                                                szs[UNIFIED_SECTION_CMDLINE]);
-                buffer = xasprintf("%ls%ls%ls", strempty(tmp), isempty(tmp) ? u"" : u" ", extra16);
-        }
+                if (ret_cmdline && szs[UNIFIED_SECTION_CMDLINE] > 0) {
+                        _cleanup_free_ char16_t *tmp = TAKE_PTR(cmdline),
+                                                *extra16 = xstrn8_to_16((char *)loaded_addon->ImageBase + addrs[UNIFIED_SECTION_CMDLINE],
+                                                                        szs[UNIFIED_SECTION_CMDLINE]);
+                        cmdline = xasprintf("%ls%ls%ls", strempty(tmp), isempty(tmp) ? u"" : u" ", extra16);
+                }
 
-        mangle_stub_cmdline(buffer);
+                if (ret_dt_bases && szs[UNIFIED_SECTION_DTB] > 0) {
+                        dt_sizes = xrealloc(dt_sizes,
+                                            n_dt * sizeof(size_t),
+                                            (n_dt + 1)  * sizeof(size_t));
+                        dt_sizes[n_dt] = szs[UNIFIED_SECTION_DTB];
 
-        if (!isempty(buffer)) {
-                _cleanup_free_ char16_t *tmp = TAKE_PTR(*cmdline_append);
-                bool m = false;
+                        dt_bases = xrealloc(dt_bases,
+                                            n_dt * sizeof(EFI_PHYSICAL_ADDRESS),
+                                            (n_dt + 1) * sizeof(EFI_PHYSICAL_ADDRESS));
+                        dt_bases[n_dt] = xmemdup(
+                                PHYSICAL_ADDRESS_TO_POINTER(POINTER_TO_PHYSICAL_ADDRESS(loaded_addon->ImageBase) + addrs[UNIFIED_SECTION_DTB]),
+                                dt_sizes[n_dt]);
```

Comment:
This should use `void*` for calculations as it's the type of `dt_bases`.

Reply:
What do you mean precisely?

Reply:
The type of `dt_bases` is `void*`, so why go through `EFI_PHYSICAL_ADDRESS`?
```suggestion
                        dt_bases = xrealloc(dt_bases,
                                            n_dt * sizeof(void*),
                                            (n_dt + 1) * sizeof(void*));
                        dt_bases[n_dt] = xmemdup((uint8_t*)loaded_addon->ImageBase + addrs[UNIFIED_SECTION_DTB], dt_sizes[n_dt]);
```

Reply:
ah in that sense - no reason, fixed

Reply:
You missed one `sizeof(EFI_PHYSICAL_ADDRESS)`.

Reply:
whops